### PR TITLE
curvebs : support spdk over nvme

### PR DESCRIPTION
**Pull request by xu-chaojie**
### What problem does this PR solve?

Issue Number: #1773  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/chunkserver/datastore/file_pool.cpp [hunk 1]
```diff
@@ -279,7 +279,7 @@ bool FilePool::CleanChunk(uint64_t chunkid, bool onlyMarked) {
         char* buffer = writeBuffer_.get();
 
         while (nwrite < ntotal) {
-            nbytes = fsptr_->Write(fd, buffer, nwrite,
+            nbytes = fsptr_->WriteZeroIfSupport(fd, buffer, nwrite,
```
Comment (skypexu) [inline src/chunkserver/datastore/file_pool.cpp:282]: 如何警告配置错误?
如果poolOpt_.bytesPerWrite 不是512的倍数，那么底层会低效率，pfs nvme 不得不读然后写合并

File: src/chunkserver/chunkserver.cpp [hunk 2]
```diff
@@ -455,6 +636,13 @@ int ChunkServer::Run(int argc, char** argv) {
         << "Failed to shutdown file pool clean worker.";
     concurrentapply.Stop();
 
+#ifdef WITH_SPDK
+    if (fs::FileSystemType::PFS == lfsOption.type) {
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:640]: Why do you need WITH_SPDK, and you need to judge if (fs::FileSystemType::PFS == lfsOption.type)?

File: src/chunkserver/chunkserver.cpp [hunk 3]
```diff
@@ -75,8 +90,127 @@ DEFINE_string(walFilePoolDir, "./0/", "WAL filepool location");
...
+    case PFS_TRACE_VERB:
+    default:
+        glevel = google::GLOG_INFO;
+        break;
+    }
+    google::LogMessage(file, line, glevel).stream() << buf;
+}
+
+void
+glog_spdk_func(int level, const char *a_file, const int a_line,
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:138]: A lot of glog_ related functions are implemented in chunkserver.cpp, but they are not used in this file. It is recommended to define them in a new file.

File: src/chunkserver/chunkserver.cpp [hunk 4]
```diff
@@ -75,8 +90,127 @@ DEFINE_string(walFilePoolDir, "./0/", "WAL filepool location");
...
+    }
+    google::LogMessage("<dpdk>", 0, glevel).stream().write(buf, size);
+    if (level == RTE_LOG_EMERG)
+        abort();
+    return size;
+}
+
+void* dpdk_mem_allocate(size_t align, size_t sz) {
+    /* rte_malloc seems fast enough, otherwise we need to use mempool */
+    return rte_malloc("iobuf", sz, align);
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:206]: Is rte_malloc a new memory allocation library?

File: src/chunkserver/datastore/chunkserver_chunkfile.h [hunk 5]
```diff
@@ -364,7 +369,7 @@ class CSChunkFile {
     }
 
     inline int SyncData() {
-        return lfs_->Sync(fd_);
+        return lfs_->Fdatasync(fd_);
```
Comment (ilixiaocui) [inline src/chunkserver/datastore/chunkserver_chunkfile.h:372]: Has this function ever been called? Generally, specify the type of flash disk when Open?

File: src/chunkserver/datastore/chunkserver_datastore.h [hunk 6]
```diff
@@ -173,6 +173,19 @@ class CSDataStore {
      */
     virtual CSErrorCode DeleteSnapshotChunkOrCorrectSn(
         ChunkID id, SequenceNum correctedSn);
+
+    // Deprecated, only use for unit & integration test
```
Comment (ilixiaocui) [inline src/chunkserver/datastore/chunkserver_datastore.h:177]: Inconsistent comment format and context.

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp [hunk 7]
```diff
@@ -22,17 +22,18 @@
...
 namespace curve {
 namespace chunkserver {
 CurveFilesystemAdaptor::CurveFilesystemAdaptor(
-                                std::shared_ptr<FilePool> chunkFilePool,
+                                std::shared_ptr<FilePool> filePool,
                                 std::shared_ptr<LocalFileSystem> lfs) {
     lfs_ = lfs;
-    chunkFilePool_ = chunkFilePool;
-    uint64_t metapageSize = chunkFilePool->GetFilePoolOpt().metaPageSize;
+    filePool_ = filePool;
```
Comment (ilixiaocui) [inline src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:35]: Why did you change the name of chunkfilepool to filepool?

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp [hunk 8]
```diff
@@ -47,28 +48,14 @@ CurveFilesystemAdaptor::~CurveFilesystemAdaptor() {
...
-    LOG(INFO) << "release raftsnapshot filesystem adaptor!";
+    LOG(INFO) << "release CurveFilesystemAdaptor!";
 }
 
 braft::FileAdaptor* CurveFilesystemAdaptor::open(const std::string& path,
                     int oflag, const ::google::protobuf::Message* file_meta,
                     butil::File::Error* e) {
     (void) file_meta;
 
-    static std::once_flag local_s_check_cloexec_once;
```
Comment (ilixiaocui) [inline src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:58]: Why was this part of the code deleted?

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp [hunk 9]
```diff
@@ -176,14 +160,108 @@ bool CurveFilesystemAdaptor::rename(const std::string& old_path,
                                             const std::string& new_path) {
     if (!NeedFilter(new_path) && lfs_->FileExists(new_path)) {
         // chunkfilepool内部会检查path对应文件合法性，如果不符合就直接删除
-        chunkFilePool_->RecycleFile(new_path);
+        filePool_->RecycleFile(new_path);
     }
     return lfs_->Rename(old_path, new_path) == 0;
 }
 
+bool CurveFilesystemAdaptor::link(const std::string& old_path,
```
Comment (ilixiaocui) [inline src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:168]: Why was this part of the code moved out of the sanpshot directory?

File: src/chunkserver/raftlog/curve_segment.cpp [hunk 10]
```diff
@@ -604,12 +603,16 @@ int CurveSegment::close(bool will_sync) {
     if (_last_index > _first_index) {
         if (FLAGS_raftSyncSegments && will_sync &&
                                 !FLAGS_enableWalDirectWrite) {
-            ret = braft::raft_fsync(_fd);
+            if (braft::FLAGS_raft_use_fsync_rather_than_fdatasync) {
```
Comment (ilixiaocui) [inline src/chunkserver/raftlog/curve_segment.cpp:606]: This means that fdatasync can be used when using filepool, right? Why add another flag? Don't use the walpool related configuration in the configuration file directly.

File: src/fs/pfs_filesystem_impl.cpp [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+    ss << "{";
+    for (size_t i = 0; i < nvec; ++i) {
+        ss << "[" << uintptr_t(vec[i].iov_base) << ","
+           << vec[i].iov_len << "]";
+    }
+    ss << "}";
+    return ss.str();
+}
+
+ssize_t preadv_dispatch(int fd, const struct iovec *vector,
```
Comment (ilixiaocui) [inline src/fs/pfs_filesystem_impl.cpp:59]: Explain the function of this function

File: src/fs/pfs_filesystem_impl.cpp [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+                       << ", nvec: " << nvec
+                       << ", offset: " << offset
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return preadv(fd, vector, nvec, offset);
+}
+
+ssize_t pwritev_dispatch(int fd, const struct iovec *vector,
```
Comment (ilixiaocui) [inline src/fs/pfs_filesystem_impl.cpp:75]: Explain the function of this function

File: src/fs/pfs_filesystem_impl.cpp [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+                       << ", nvec: " << nvec
+                       << ", offset: " << offset
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return pwritev(fd, vector, nvec, offset);
+}
+
+ssize_t readv_dispatch(int fd, const struct iovec *vector, int nvec) {
```
Comment (ilixiaocui) [inline src/fs/pfs_filesystem_impl.cpp:92]: ditto

File: src/fs/pfs_filesystem_impl.cpp [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+                       << ", fd: " << fd
+                       << ", nvec: " << nvec
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return readv(fd, vector, nvec);
+}
+
+ssize_t writev_dispatch(int fd, const struct iovec *vector, int nvec) {
```
Comment (ilixiaocui) [inline src/fs/pfs_filesystem_impl.cpp:106]: ditto

Comment (ilixiaocui) [reply src/fs/pfs_filesystem_impl.cpp:106]: What's the diffrence between pwritev_dispatch and writev_dispatch.

File: src/tools/curve_format_main.cpp [hunk 15]
```diff
@@ -86,11 +107,119 @@ DEFINE_bool(needWriteZero,
...
+    "0000:c2:00.0n1",
+    "pfs pbd name, only for pfs");
+
+DEFINE_int32(pfs_host_id,
+    2,
+    "pfs host id, only for pfs");
+
+#ifdef WITH_SPDK
+void
+glog_pfs_func(int level, const char *file, const char *func, int line,
```
Comment (ilixiaocui) [inline src/tools/curve_format_main.cpp:128]: This part of the function is also defined in chunkserver.cpp.

File: src/tools/curve_format_main.cpp [hunk 16]
```diff
@@ -177,12 +306,52 @@ int AllocateFiles(AllocateStruct* allocatestruct) {
...
+    FILE *dpdk_log_stream = NULL;
+    cookie_io_functions_t io_funcs;
+    memset(&io_funcs, 0, sizeof(io_funcs));
+#endif
+
     google::ParseCommandLineFlags(&argc, &argv, false);
     google::InitGoogleLogging(argv[0]);
 
+#ifdef WITH_SPDK
+    io_funcs.write = glog_dpdk_log_func;
```
Comment (ilixiaocui) [inline src/tools/curve_format_main.cpp:319]: What is the difference between the format of spdk and the format of ext4？

File: WORKSPACE [hunk 17]
```diff
@@ -227,3 +225,55 @@ http_archive(
 )
 load("@hedron_compile_commands//:workspace_setup.bzl", "hedron_compile_commands_setup")
 hedron_compile_commands_setup()
+
+new_local_repository(
+    name = "pfs",
+    path = "/usr/local/polarstore/pfsd/include",
+    build_file_content = """
+package(default_visibility = ["//visibility:public"])
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply WORKSPACE:233]: done

File: conf/chunkserver.conf [hunk 18]
```diff
@@ -144,6 +144,10 @@ curve.curve_file_timeout_s=30
 #
 # 是否开启使用renameat2，ext4内核3.15以后开始支持
 fs.enable_renameat2=true
+fs.type=ext4  #ext4, pfs
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply conf/chunkserver.conf:147]: done

File: conf/chunkserver.conf.example [hunk 19]
```diff
@@ -144,6 +144,10 @@ curve.curve_file_timeout_s=30
 #
 # 是否开启使用renameat2，ext4内核3.15以后开始支持
 fs.enable_renameat2=true
+fs.type=ext4  #ext4, pfs
+fs.pfs_cluster=spdk #cluster name for pfs
+fs.pfs_pbd_name=0000:81:00.0n1 #pdb name for pfs
+fs.pfs_host_id=2 #host id for pfs
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply conf/chunkserver.conf.example:150]: yes

File: mk-deb.sh [hunk 20]
```diff
@@ -487,35 +504,35 @@ if [ $? -ne 0 ]
...
-then
-    exit
-fi
-cp -r ./curve-snapshotcloneserver-nginx/docker \
-build/curve-nginx/etc/curve/nginx/
-if [ $? -ne 0 ]
-then
-    exit
-fi
+#cp -r ./curve-snapshotcloneserver-nginx/app/lib \
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply mk-deb.sh:507]: I will revert it

File: src/chunkserver/chunkserver.cpp (hunk 2 above)
Comment (xu-chaojie) [reply src/chunkserver/chunkserver.cpp:640]: done

File: src/chunkserver/chunkserver.cpp (hunk 3 above)
Comment (xu-chaojie) [reply src/chunkserver/chunkserver.cpp:138]: done

File: src/chunkserver/chunkserver.cpp (hunk 4 above)
Comment (xu-chaojie) [reply src/chunkserver/chunkserver.cpp:206]: It is dpdk memory allocation

File: src/chunkserver/datastore/chunkserver_chunkfile.h (hunk 5 above)
Comment (xu-chaojie) [reply src/chunkserver/datastore/chunkserver_chunkfile.h:372]: use sync  in the background

File: src/chunkserver/datastore/chunkserver_datastore.h (hunk 6 above)
Comment (xu-chaojie) [reply src/chunkserver/datastore/chunkserver_datastore.h:177]: done

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp (hunk 7 above)
Comment (xu-chaojie) [reply src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:35]: because it can be walpool

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp (hunk 8 above)
Comment (xu-chaojie) [reply src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:58]: we not support cloexec here

File: src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp (hunk 9 above)
Comment (xu-chaojie) [reply src/chunkserver/filesystem_adaptor/curve_filesystem_adaptor.cpp:168]: becase curve segment also should use CurveFilesystemAdaptor to access localfilesysterm.

File: src/chunkserver/raftlog/curve_segment.cpp (hunk 10 above)
Comment (xu-chaojie) [reply src/chunkserver/raftlog/curve_segment.cpp:606]: This logic is same as  braft::raft_fsync

File: src/fs/pfs_filesystem_impl.cpp (hunk 11 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:59]: It is the hook used by iobuf to dispatch io use dma or not.

File: src/fs/pfs_filesystem_impl.cpp (hunk 12 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:75]: It is the hook function used by iobuf to dispatch io use dma or not.
we add a comment here.

File: src/fs/pfs_filesystem_impl.cpp (hunk 13 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:92]: It is the hook function used by iobuf to dispatch io use dma or not.
we add a comment here.

File: src/fs/pfs_filesystem_impl.cpp (hunk 14 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:106]: pwritev_dispatch's parameter has offset feild

File: src/tools/curve_format_main.cpp (hunk 15 above)
Comment (xu-chaojie) [reply src/tools/curve_format_main.cpp]: done

File: src/tools/curve_format_main.cpp (hunk 16 above)
Comment (xu-chaojie) [reply src/tools/curve_format_main.cpp:222]: It is same as ext4

**Comment by xu-chaojie** [comment]
recheck

File: WORKSPACE (hunk 17 above)
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply WORKSPACE]: Adding an indent here will cause an error： 
image

File: src/chunkserver/raftlog/curve_segment.cpp [hunk 21]
```diff
@@ -71,26 +75,23 @@ int CurveSegment::create() {
         LOG(ERROR) << "Get segment from chunk file pool fail!";
         return -1;
     }
-    _fd = ::open(path.c_str(), O_RDWR|O_NOATIME, 0644);
-    if (_fd >= 0) {
-        butil::make_close_on_exec(_fd);
-    } else {
+    _fd = _lfs->Open(path, O_RDWR|O_NOATIME);
```
Comment (wu-hanqing) [inline src/chunkserver/raftlog/curve_segment.cpp:78]: why remove the thrid parameter `0644`?

File: src/chunkserver/spdk_hook.h [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+#include <rte_log.h>
+#include <rte_malloc.h>
+#include <pfs_spdk.h>
+#include <pfs_api.h>
+#include <pfs_trace_func.h>
+#include <pfsd.h>
+#include <pfsd_sdk.h>
+#include <pfsd_sdk_log.h>
+
+void glog_pfs_func(int level, const char *file, const char *func, int line,
```
Comment (wu-hanqing) [inline src/chunkserver/spdk_hook.h:36]: did you forget to put these functions in a namespace?

File: src/fs/fs_common.h [hunk 23]
```diff
@@ -23,14 +23,31 @@
...
 namespace fs {
 
 enum class FileSystemType {
     // SFS,
     EXT4,
+    PFS,
+    UNKOWN,
 };
 
+inline static FileSystemType StringToFileSystemType(
```
Comment (wu-hanqing) [inline src/fs/fs_common.h:38]: what's the purpose of `static` here? I think only `inline` is necessary.

File: src/fs/pfs_filesystem_impl.cpp [hunk 24]
```diff
@@ -0,0 +1,569 @@
...
+                            remainLength,
+                            offset);
+        if (ret == 0) {
+            LOG(WARNING) << "pread returns zero."
+                         << "offset: " << offset
+                         << ", length: " << remainLength;
+            break;
+        }
+        if (ret < 0) {
+            if (errno == EINTR && retryTimes < 3) {
```
Comment (wu-hanqing) [inline src/fs/pfs_filesystem_impl.cpp:401]: should you add max retry times in the config file?

File: src/fs/pfs_filesystem_impl.cpp [hunk 25]
```diff
@@ -0,0 +1,569 @@
...
+    int remainLength = length;
+    int relativeOffset = 0;
+    int retryTimes = 0;
+    while (remainLength > 0) {
+        int ret = pfs_pwrite(fd,
+                             buf + relativeOffset,
+                             remainLength,
+                             offset);
+        if (ret < 0) {
+            if (errno == EINTR && retryTimes < 3) {
```
Comment (wu-hanqing) [inline src/fs/pfs_filesystem_impl.cpp:450]: ditto

File: src/fs/pfs_filesystem_impl.cpp [hunk 26]
```diff
@@ -0,0 +1,569 @@
...
+    iofuncs.iof_readv = readv_dispatch;
+    iofuncs.iof_writev = writev_dispatch;
+    butil::iobuf::set_external_io_funcs(iofuncs);
+}
+
+std::shared_ptr<PfsFileSystemImpl> PfsFileSystemImpl::self_ = nullptr;
+std::mutex PfsFileSystemImpl::mutex_;
+
+std::shared_ptr<PfsFileSystemImpl> PfsFileSystemImpl::getInstance() {
+    std::lock_guard<std::mutex> lock(mutex_);
```
Comment (wu-hanqing) [inline src/fs/pfs_filesystem_impl.cpp:173]: `std::call_once` is more suitable

File: src/chunkserver/datastore/chunkserver_datastore.h [hunk 27]
```diff
@@ -187,6 +188,27 @@ class CSDataStore {
                                   SequenceNum sn,
                                   char * buf,
                                   off_t offset,
+                                  size_t length) {
+        butil::IOPortal iop;
+        CSErrorCode rc = ReadChunk(id, sn, &iop, offset, length);
+        iop.copy_to(buf, length, 0);
```
Comment (wu-hanqing) [inline src/chunkserver/datastore/chunkserver_datastore.h:194]: the third parameter is 0 by default, maybe you should not specify it explicitly

File: src/chunkserver/chunkserver.cpp [hunk 28]
```diff
@@ -75,6 +79,20 @@ DEFINE_string(walFilePoolDir, "./0/", "WAL filepool location");
...
+DEFINE_string(pfs_pbd_name,
+    "",
+    "pfs pbd name, only for pfs");
+DEFINE_int32(pfs_host_id,
+    -1,
+    "pfs host id, only for pfs");
+DEFINE_string(spdk_nvme_controller,
+    "",
+    "spdk nvme controller, only for pfs");
+
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply src/chunkserver/chunkserver.cpp:95]: done

File: src/chunkserver/datastore/chunkserver_chunkfile.h [hunk 29]
```diff
@@ -323,6 +324,10 @@ class CSChunkFile {
         return lfs_->Read(fd_, buf, offset + pageSize_, length);
     }
 
+    inline int readData(butil::IOPortal* buf, off_t offset, size_t length) {
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply src/chunkserver/datastore/chunkserver_chunkfile.h:327]: But it's harmless

File: src/chunkserver/raftlog/curve_segment.cpp (hunk 21 above)
Comment (xu-chaojie) [reply src/chunkserver/raftlog/curve_segment.cpp:78]: done

File: src/chunkserver/spdk_hook.h (hunk 22 above)
Comment (xu-chaojie) [reply src/chunkserver/spdk_hook.h:36]: no, I think we should put this in global namespace

File: src/fs/fs_common.h (hunk 23 above)
Comment (xu-chaojie) [reply src/fs/fs_common.h:38]: done

File: src/fs/pfs_filesystem_impl.cpp (hunk 24 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:401]: ext4 filesystem do like this

File: src/fs/pfs_filesystem_impl.cpp (hunk 26 above)
Comment (xu-chaojie) [reply src/fs/pfs_filesystem_impl.cpp:173]: it is not nessary

File: src/chunkserver/datastore/chunkserver_datastore.h (hunk 27 above)
Comment (xu-chaojie) [reply src/chunkserver/datastore/chunkserver_datastore.h:194]: But it's harmless

**Review by wu-hanqing — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]
